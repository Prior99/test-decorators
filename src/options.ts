export interface TestOptions<Params> {
    name?: string | ((params?: Params) => string)
    params?: Params[],
    only?: boolean
}

export type TestOptionsInput<Params> = TestOptions<Params> | string

export interface PreparedTestOptions<Params> {
    name: (params?: Params) => string
    params?: Params[],
    only: boolean,
}

export function parseTestOptions<Params>(options?: TestOptionsInput<Params>): PreparedTestOptions<Params> {
    if (typeof options === "string") {
        return {
            name: () => options,
            only: false,
        }
    }
    if (typeof options !== "object") {
        throw new Error("Test options need to be either a string or an object.")
    }
    if (options === null) {
        throw new Error("Test options can not be null.")
    }
    const { name: originalName, params, only } = options
    if (typeof originalName !== "function" && typeof originalName !== "string" && originalName !== undefined) {
        throw new Error("Option 'name' needs to be either a string or a function or undefined.")
    }
    let name: () => string
    if (typeof originalName === "function") {
        name = originalName
    }
    if (typeof originalName === "string") {
        name = () => originalName
    }
    if (params !== undefined && !Array.isArray(params)) {
        throw new Error("Option 'params' needs to be an array.")
    }
    return { name, params, only }
}

export function isTestOptionsInput<Params>(input: any): input is TestOptionsInput<Params> {
    const validKeys = ["name", "params", "only"]
    if (typeof input === "string") { return true }
    if (input.constructor !== Object) { return false }
    if (Object.keys(input).some(key => !validKeys.includes(key))) { return false }
    return true
}

export interface SuiteOptions {
    name?: string
    only?: boolean
}

export type SuiteOptionsInput = SuiteOptions | string

export interface PreparedSuiteOptions {
    name: string
    only: boolean
}

export function parseSuiteOptions(options?: SuiteOptionsInput): PreparedSuiteOptions {
    if (typeof options === "string") {
        return {
            name: options,
            only: false,
        }
    }
    if (typeof options !== "object") {
        throw new Error("Suite options need to be either a string or an object.")
    }
    if (options === null) {
        throw new Error("Suite options can not be null.")
    }
    const { name, only } = options
    if (name !== undefined && typeof name !== "string") {
        throw new Error("Option 'name' needs to be either a string or a function.")
    }
    return { name, only }
}

export function isSuiteOptionsInput(input: any): input is SuiteOptionsInput {
    if (typeof input === "string") { return true }
    if (typeof input === "function") { return false }
    return true
}
