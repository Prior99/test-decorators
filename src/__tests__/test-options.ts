import { parseTestOptions, parseSuiteOptions, isTestOptionsInput, isSuiteOptionsInput } from "../options"

describe("`parseTestOptions`", () => {
    test("with only a name", () => {
        const options = parseTestOptions("the name of some test")
        expect(options).toMatchSnapshot()
        expect(options.name()).toBe("the name of some test")
    })

    test("with `undefined`", () => {
        expect(() => parseTestOptions(undefined)).toThrowErrorMatchingSnapshot()
    })

    test("with `null`", () => {
        expect(() => parseTestOptions(null)).toThrowErrorMatchingSnapshot() // tslint:disable-line
    })

    test("with a number", () => {
        expect(() => parseTestOptions(9 as any)).toThrowErrorMatchingSnapshot()
    })

    test("with the name being a string", () => {
        const options = parseTestOptions({
            name: "some test name",
        })
        expect(options).toMatchSnapshot()
        expect(options.name()).toBe("some test name")
    })

    test("with the name being a function", () => {
        const mock = jest.fn()
        const options = parseTestOptions({
            name: () => {
                mock()
                return "some test name in a function"
            },
        })
        expect(options).toMatchSnapshot()
        expect(options.name()).toBe("some test name in a function")
        expect(mock).toHaveBeenCalled()
    })

    test("with the name being undefined", () => {
        const options = parseTestOptions({})
        expect(options).toMatchSnapshot()
    })

    test("with the name being a number", () => {
        expect(() => parseTestOptions({ name: 9 } as any)).toThrowErrorMatchingSnapshot()
    })

    test("with the params being an object which is not an array", () => {
        expect(() => parseTestOptions({ params: {} } as any)).toThrowErrorMatchingSnapshot()
    })

    test("with the params being an array", () => {
        const options = parseTestOptions({
            params: [
                { a: 1 },
                { a: 2 },
                { a: 3 },
            ],
        })
        expect(options).toMatchSnapshot()
    })

    test("with only being set to `true`", () => {
        const options = parseTestOptions({
            only: true,
        })
        expect(options).toMatchSnapshot()
    })

    test("with a full configuration", () => {
        const options = parseTestOptions({
            only: true,
            name: "yet another test name",
            params: [1, 2, 3],
        })
        expect(options.name()).toBe("yet another test name")
        expect(options).toMatchSnapshot()
    })
})

describe("`parseSuiteOptions`", () => {
    test("with only a name", () => {
        const options = parseSuiteOptions("the name of some suite")
        expect(options).toMatchSnapshot()
    })

    test("with `undefined`", () => {
        expect(() => parseSuiteOptions(undefined)).toThrowErrorMatchingSnapshot()
    })

    test("with `null`", () => {
        expect(() => parseSuiteOptions(null)).toThrowErrorMatchingSnapshot() // tslint:disable-line
    })

    test("with a number", () => {
        expect(() => parseSuiteOptions(9 as any)).toThrowErrorMatchingSnapshot()
    })

    test("with the name specified", () => {
        const options = parseSuiteOptions({
            name: "some suite name",
        })
        expect(options).toMatchSnapshot()
    })

    test("with the name being undefined", () => {
        const options = parseSuiteOptions({})
        expect(options).toMatchSnapshot()
    })

    test("with the name being a number", () => {
        expect(() => parseSuiteOptions({ name: 9 } as any)).toThrowErrorMatchingSnapshot()
    })

    test("with only being set to `true`", () => {
        const options = parseSuiteOptions({
            only: true,
        })
        expect(options).toMatchSnapshot()
    })

    test("with a full configuration", () => {
        const options = parseSuiteOptions({
            only: true,
            name: "another suite name",
        })
        expect(options).toMatchSnapshot()
    })
})

describe("`isTestOptionsInput`", () => {
    test("doesn't detect a class as options", () => {
        expect(isTestOptionsInput(class A{})).toBe(false)
    })

    test("doesn't detect any instance as options", () => {
        class A {
            public name = "test"
            public only = true
            public params = [1, 2, 3]
        }
        expect(isTestOptionsInput(new A())).toBe(false)
    })

    test("doesn't detect a method as options", () => {
        class A {
            public someMethod() { return }
        }
        const instance = new A()
        expect(isTestOptionsInput(instance.someMethod)).toBe(false)
        expect(isTestOptionsInput(A.prototype.someMethod)).toBe(false)
    })

    test("doesn't detect a function as options", () => {
        expect(isTestOptionsInput(() => { return })).toBe(false)
    })

    test("doesn't detect any unrelated object as options", () => {
        expect(isTestOptionsInput({ someUnknownKey: 10 })).toBe(false)
        expect(isTestOptionsInput({
            someUnknownKey: "test",
            only: true,
            name: "some name",
        })).toBe(false)
    })

    test("doesn't detect null as options", () => {
        expect(isTestOptionsInput(null)).toBe(false) // tslint:disable-line
    })

    test("doesn't detect undefined as options", () => {
        expect(isTestOptionsInput(undefined)).toBe(false)
    })

    test("detects a string as options", () => {
        expect(isTestOptionsInput("some test name")).toBe(true)
    })

    test("detects empty options as options", () => {
        expect(isTestOptionsInput({})).toBe(true)
    })

    test("detects well-defined options as options", () => {
        expect(isTestOptionsInput({ only: true, name: "a test name", params: [] })).toBe(true)
    })
})

describe("`isSuiteOptionsInput`", () => {
    test("doesn't detect a class as options", () => {
        expect(isSuiteOptionsInput(class A{})).toBe(false)
    })

    test("doesn't detect any instance as options", () => {
        class A {
            public name = "test"
            public only = true
            public params = [1, 2, 3]
        }
        expect(isSuiteOptionsInput(new A())).toBe(false)
    })

    test("doesn't detect a method as options", () => {
        class A {
            public someMethod() { return }
        }
        const instance = new A()
        expect(isSuiteOptionsInput(instance.someMethod)).toBe(false)
        expect(isSuiteOptionsInput(A.prototype.someMethod)).toBe(false)
    })

    test("doesn't detect a function as options", () => {
        expect(isSuiteOptionsInput(() => { return })).toBe(false)
    })

    test("doesn't detect any unrelated object as options", () => {
        expect(isSuiteOptionsInput({ someUnknownKey: 10 })).toBe(false)
        expect(isSuiteOptionsInput({
            someUnknownKey: "test",
            only: true,
            name: "some name",
        })).toBe(false)
    })

    test("doesn't detect null as options", () => {
        expect(isSuiteOptionsInput(null)).toBe(false) // tslint:disable-line
    })

    test("doesn't detect undefined as options", () => {
        expect(isSuiteOptionsInput(undefined)).toBe(false)
    })

    test("detects a string as options", () => {
        expect(isSuiteOptionsInput("some suite name")).toBe(true)
    })

    test("detects empty options as options", () => {
        expect(isSuiteOptionsInput({})).toBe(true)
    })

    test("detects well-defined options as options", () => {
        expect(isSuiteOptionsInput({ only: true, name: "a suite name", params: [] })).toBe(true)
    })
})
