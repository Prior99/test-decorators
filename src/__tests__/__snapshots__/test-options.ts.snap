// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`parseSuiteOptions\` with \`null\` 1`] = `"Suite options can not be null."`;

exports[`\`parseSuiteOptions\` with \`undefined\` 1`] = `"Suite options need to be either a string or an object."`;

exports[`\`parseSuiteOptions\` with a full configuration 1`] = `
Object {
  "name": "another suite name",
  "only": true,
}
`;

exports[`\`parseSuiteOptions\` with a number 1`] = `"Suite options need to be either a string or an object."`;

exports[`\`parseSuiteOptions\` with only a name 1`] = `
Object {
  "name": "the name of some suite",
  "only": false,
}
`;

exports[`\`parseSuiteOptions\` with only being set to \`true\` 1`] = `
Object {
  "name": undefined,
  "only": true,
}
`;

exports[`\`parseSuiteOptions\` with the name being a number 1`] = `"Option 'name' needs to be either a string or a function."`;

exports[`\`parseSuiteOptions\` with the name being undefined 1`] = `
Object {
  "name": undefined,
  "only": undefined,
}
`;

exports[`\`parseSuiteOptions\` with the name specified 1`] = `
Object {
  "name": "some suite name",
  "only": undefined,
}
`;

exports[`\`parseTestOptions\` with \`null\` 1`] = `"Test options can not be null."`;

exports[`\`parseTestOptions\` with \`undefined\` 1`] = `"Test options need to be either a string or an object."`;

exports[`\`parseTestOptions\` with a full configuration 1`] = `
Object {
  "name": [Function],
  "only": true,
  "params": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`\`parseTestOptions\` with a number 1`] = `"Test options need to be either a string or an object."`;

exports[`\`parseTestOptions\` with only a name 1`] = `
Object {
  "name": [Function],
  "only": false,
}
`;

exports[`\`parseTestOptions\` with only being set to \`true\` 1`] = `
Object {
  "name": undefined,
  "only": true,
  "params": undefined,
}
`;

exports[`\`parseTestOptions\` with the name being a function 1`] = `
Object {
  "name": [Function],
  "only": undefined,
  "params": undefined,
}
`;

exports[`\`parseTestOptions\` with the name being a number 1`] = `"Option 'name' needs to be either a string or a function or undefined."`;

exports[`\`parseTestOptions\` with the name being a string 1`] = `
Object {
  "name": [Function],
  "only": undefined,
  "params": undefined,
}
`;

exports[`\`parseTestOptions\` with the name being undefined 1`] = `
Object {
  "name": undefined,
  "only": undefined,
  "params": undefined,
}
`;

exports[`\`parseTestOptions\` with the params being an array 1`] = `
Object {
  "name": undefined,
  "only": undefined,
  "params": Array [
    Object {
      "a": 1,
    },
    Object {
      "a": 2,
    },
    Object {
      "a": 3,
    },
  ],
}
`;

exports[`\`parseTestOptions\` with the params being an object which is not an array 1`] = `"Option 'params' needs to be an array."`;
